Feature/debian installer

julsemaan opened the pull request:
# Description
Adds tooling to create an ISO installer for PacketFence based on Debian 11.2

# Impacts
Nothing, just adds tooling

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [N/A] Add unit tests
- [ ] Add acceptance tests (TestLink)

julsemaan commented:
This still needs to be integrated in our release pipeline, I tested the ISO manually and I get through the configurator

julsemaan commented:
This should be ready, @nqb, can you review?

You can download the ISO from here if you want to test:
https://sourceforge.net/projects/packetfence/files/PacketFence%20ISO/feature-debian-installer/

nqb commented on .gitlab-ci.yml:
```diff
@@ -33,10 +33,12 @@ variables:
   TESTDIR: t/venom
   UPLOAD_DIR: $CILIBDIR/upload
   SF_REPO_URL: https://sourceforge.net/projects/packetfence/files/PacketFence%20ZEN
+  SF_REPO_URL: https://sourceforge.net/projects/packetfence/files/PacketFence%20ISO
```
It will override previous line (same variable name)

nqb commented on ci/debian-installer/build-and-upload.sh:
```diff
@@ -0,0 +1,35 @@
+
+mkdir -p ${SF_RESULT_DIR}
+
+# Manifest need to be generate by hand because we modify OVF during last step
```
Can be removed

nqb commented on ci/debian-installer/preseed.cfg.tmpl:
```diff
@@ -0,0 +1,507 @@
+#d-i debian-installer/locale string en_GB.UTF-8
+# Optionally specify additional locales to be generated.
+#d-i localechooser/supported-locales multiselect en_US.UTF-8, nl_NL.UTF-8
+d-i localechooser/supported-locales multiselect en_US.UTF-8, \
+                                                es_ES.UTF-8, \
+                                                fr_FR.UTF-8, \
+                                                fr_CA.UTF-8, \
+                                                de_DE.UTF-8, \
+                                                he_IL.UTF-8, \
+                                                it_IT.UTF-8, \
+                                                nb_NO.UTF-8, \
+                                                nl_NL.UTF-8, \
+                                                pl_PL.UTF-8, \
+                                                pt_BR.UTF-8
+
```
Can be removed, see https://github.com/inverse-inc/packetfence/commit/3298704a27b444eb3319debe8bec2ea987c96ff1

nqb commented on ci/debian-installer/create-debian-installer.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
This script will not fail in case of error.

nqb commented on .gitlab-ci.yml:
```diff
@@ -353,6 +363,17 @@ variables:
+  stage: build_pf_img
+  environment:
+    name: sourceforge
+    url: ${SF_REPO_URL}/${CI_COMMIT_REF_SLUG}
```
This URL will be wrong for a release (tag) because we use ${CI_COMMIT_TAG} later.
This has a minimal effect

julsemaan replied:
I added logic to use the tag for releases and the ref for devel related builds

nqb reviewed (changes requested):
Minor changes but it looks good. I didn't test ZEN ISO.

General question: I compare `preseed.cfg.tmpl` with `preseed.cfg` used for ZEN. I see password is not set but I think it's to let user define it.

julsemaan commented:
> Minor changes but it looks good. I didn't test ZEN ISO.
> 
> General question: I compare `preseed.cfg.tmpl` with `preseed.cfg` used for ZEN. I see password is not set but I think it's to let user define it.

That is correct, since the install is interactive in this case, I'm letting users pick their own root password instead of having to change it post-install.

I addressed your comments and rebuilt the ISO and tested it in my lab and it works.
If you want to test it, its available on sourceforge

nqb commented:
I tested ISO with success. I have one remaining comment: should we not disable predictable network names ? 

It's the case in ZEN image but it's more to simplify provisioning. However, I'm thinking about cluster installation where you can have different network names and you will need to disable it anyways.

Option to add to preseed should be: https://github.com/inverse-inc/packetfence/commit/78f903e917dd304c4800efd2ff0c00054fc1caca#diff-2e9669c2d68bf0daf463f91cf709f56da70b2ba792af71b0e5728f0f7ea7c670

nqb commented:
We will need to update packetfence.org to mention new image and add a section in Installation guide.

nqb commented:
Could you disable AppArmor (like in ZEN image) ? I think you just need to add `apparmor=0` after `d-i debian-installer/add-kernel-opts string`

julsemaan commented:
I noticed that the 'Graphical install' option ignores the preseed. I'm working on removing that option from the ISO too

julsemaan commented:
> I tested ISO with success. I have one remaining comment: should we not disable predictable network names ?
> 
> It's the case in ZEN image but it's more to simplify provisioning. However, I'm thinking about cluster installation where you can have different network names and you will need to disable it anyways.
> 
> Option to add to preseed should be: [78f903e#diff-2e9669c2d68bf0daf463f91cf709f56da70b2ba792af71b0e5728f0f7ea7c670](https://github.com/inverse-inc/packetfence/commit/78f903e917dd304c4800efd2ff0c00054fc1caca#diff-2e9669c2d68bf0daf463f91cf709f56da70b2ba792af71b0e5728f0f7ea7c670)

The downside with this is that all the network configuration you do during the installer is based on the 'non-predictable' interface name so when you boot into the OS, eth0 is down and not configured in /etc/network/interfaces.

nqb commented:
This was the issue we had with ZEN image. In that case, we can:
* keep things as is
* pass options directly to the installer to start install with correct options

julsemaan commented:
The installer accepts `net.ifnames=0` and combined with it in the preseed kernel options we're all good. So the interfaces can be configured using the installer and will keep predictable names both in the installer and after the install

I retested an ISO and it works but haven't republished since I'd need to realter the .gitlab-ci file. I can provide an ISO to test if needed.

Once we're happy with this, I'll write the steps in our install guide to get this installed

nqb commented:
Tested again, work as expected.
@julsemaan, I will let you adjust documentation.

julsemaan commented:
> Tested again, work as expected. @julsemaan, I will let you adjust documentation.

Just did the instructions, went pretty generic so that it survives over time

@extrafu, if you want to also check the instructions for verbiage and grammar

extrafu commented on docs/installation/installation.asciidoc:
```diff
@@ -54,6 +54,27 @@ Make sure that there is only one virtual network card created, and also make sur
+
+This setup has been tested using VMware ESXi, Proxmox VE and VirtualBox but will also work with any hypervisor PacketFence supports as well as bare-metal servers.
+
+You will need a virtual machine or server with 12 GB of RAM dedicated to machine as well as 2 CPUs. Make sure you also have a hard-drive of at least 100GB.
```
Make sure you also have a hard-drive of at least 100GB. => Make sure you allocate at least 100GB of disk space for PacketFence.

extrafu commented on docs/installation/installation.asciidoc:
```diff
@@ -54,6 +54,27 @@ Make sure that there is only one virtual network card created, and also make sur
+
+==== Installing the ISO to a bare-metal server
+
+First, make sure your server follows the specifications above and then burn the ISO onto a DVD and boot your server using the DVD.
```
Would that work from a USB stick too? If so, it's worth mentioning!

julsemaan replied:
Done! I also talked about putting it on a tape-cassette and floppy disks

extrafu commented:
Minor nits mentioned

julsemaan commented:
@nqb, all yours for the final review and merge
